File: app/src/main/java/com/sysu/edu/extra/ui/SettingFragment.java
package com.sysu.edu.extra.ui;

import android.content.ComponentName;
import android.content.pm.PackageManager;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.preference.PreferenceFragmentCompat;

import com.sysu.edu.R;
import com.sysu.edu.preference.MenuPreference;
import com.sysu.edu.preference.Theme;

import java.util.Objects;

public class SettingFragment extends PreferenceFragmentCompat {

    @Override
    public void onCreatePreferences(@Nullable Bundle savedInstanceState, @Nullable String rootKey) {
        setPreferencesFromResource(R.xml.setting, rootKey);
        Theme th = new Theme(requireActivity());
        th.setTheme();
        ((MenuPreference) Objects.requireNonNull(findPreference("theme"))).setOnPreferenceChangeListener((preference, newValue) -> {
            requireActivity().recreate();
            return true;
        });
        ((MenuPreference) Objects.requireNonNull(findPreference("icon_theme"))).setOnPreferenceChangeListener((preference, newValue) -> {
            PackageManager pm = requireActivity().getPackageManager();
            String pkg = requireContext().getPackageName();
            pm.setComponentEnabledSetting(new ComponentName(requireActivity().getBaseContext(), pkg + ".MainActivityDefault"), (new int[]{2, 2, 1})[Integer.parseInt((String) newValue)], PackageManager.DONT_KILL_APP);
            pm.setComponentEnabledSetting(new ComponentName(requireActivity().getBaseContext(), pkg + ".MainActivityLight"), (new int[]{1, 2, 2})[Integer.parseInt((String) newValue)], PackageManager.DONT_KILL_APP);
            pm.setComponentEnabledSetting(new ComponentName(requireActivity().getBaseContext(), pkg + ".MainActivityDark"), (new int[]{2, 1, 2})[Integer.parseInt((String) newValue)], PackageManager.DONT_KILL_APP);
            return true;
        });
        ((MenuPreference) Objects.requireNonNull(findPreference("language"))).setOnPreferenceChangeListener((preference, v) -> {
                    //Language.setLanguage(requireActivity());
                    requireActivity().recreate();
                    return true;
                }
        );
        ((MenuPreference) Objects.requireNonNull(findPreference("fontSize"))).setOnPreferenceChangeListener((preference, v) -> {
                    //Language.setLanguage(requireActivity());
                    requireActivity().recreate();

                    return true;
                }
        );
    }
}

File: app/src/main/java/com/sysu/edu/extra/SettingFragment.kt
package com.sysu.edu.extra

import android.content.ComponentName
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.sysu.edu.R
import com.sysu.edu.browser.BrowserPreference
import com.sysu.edu.preference.MenuPreference
import rikka.material.preference.MaterialSwitchPreference

class SettingFragment : PreferenceFragmentCompat() {
	override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
		setPreferencesFromResource(R.xml.settings, rootKey)
		((findPreference("theme")) as MenuPreference?)?.setOnPreferenceChangeListener { _: Preference?, _: Any? ->
			requireActivity().recreate()
			true
		}
		((findPreference("icon_theme")) as MenuPreference?)?.setOnPreferenceChangeListener { _: Preference?, newValue: Any? ->
			val pkg = requireContext().packageName
			with(requireActivity().packageManager) {
				setComponentEnabledSetting(ComponentName(requireActivity().baseContext, "$pkg.MainActivity"), PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP)
				setComponentEnabledSetting(ComponentName(requireActivity().baseContext, "$pkg.MainActivityLight"), (intArrayOf(1, 2, 2))[(newValue as String?)!!.toInt()], PackageManager.DONT_KILL_APP)
				setComponentEnabledSetting(ComponentName(requireActivity().baseContext, "$pkg.MainActivityDark"), (intArrayOf(2, 1, 2))[newValue.toInt()], PackageManager.DONT_KILL_APP)
				setComponentEnabledSetting(ComponentName(requireActivity().baseContext, "$pkg.MainActivityDefault"), (intArrayOf(2, 2, 1))[newValue.toInt()], PackageManager.DONT_KILL_APP)
			}
			true
		}
		((findPreference("language")) as MenuPreference?)?.setOnPreferenceChangeListener { _: Preference?, _: Any? ->
			requireActivity().recreate()
			true
		}
		((findPreference("fontSize")) as MenuPreference?)?.setOnPreferenceChangeListener { _: Preference?, _: Any? ->
			requireActivity().recreate()
			true
		}
		val browserPreference = BrowserPreference(requireContext())
		((findPreference("image_blocked")) as MaterialSwitchPreference?)?.setChecked(browserPreference.isImageBlocked)
		((findPreference("image_blocked")) as MaterialSwitchPreference?)?.setOnPreferenceChangeListener { _: Preference?, newValue: Any? ->
			browserPreference.isImageBlocked = newValue as Boolean
			true
		}
		((findPreference("js_enabled")) as MaterialSwitchPreference?)?.setChecked(browserPreference.isJSEnabled)
		((findPreference("js_enabled")) as MaterialSwitchPreference?)?.setOnPreferenceChangeListener { _: Preference?, newValue: Any? ->
			browserPreference.isJSEnabled = newValue as Boolean
			true
		}
		((findPreference("save_mobile_data_mode")) as MaterialSwitchPreference?)?.setChecked(browserPreference.isSaveMobileDataMode)
		((findPreference("save_mobile_data_mode")) as MaterialSwitchPreference?)?.setOnPreferenceChangeListener { _: Preference?, newValue: Any? ->
			browserPreference.isSaveMobileDataMode = newValue as Boolean
			true
		}
		((findPreference("privacy_mode")) as MaterialSwitchPreference?)?.setChecked(browserPreference.isPrivacyMode)
		((findPreference("privacy_mode")) as MaterialSwitchPreference?)?.setOnPreferenceChangeListener { _: Preference?, newValue: Any? ->
			browserPreference.isPrivacyMode = newValue as Boolean
			true
		}
		((findPreference("mobile_mode")) as MaterialSwitchPreference?)?.setChecked(browserPreference.isSaveMobileDataMode)
		((findPreference("mobile_mode")) as MaterialSwitchPreference?)?.setOnPreferenceChangeListener { _: Preference?, newValue: Any? ->
			browserPreference.isSaveMobileDataMode = newValue as Boolean
			true
		}
		(findPreference("developer_mode") as Preference?)?.isVisible = PreferenceManager.getDefaultSharedPreferences(requireContext()).getBoolean("developer_mode", false)
	}
}
